Improve export operation code

ggalmazor opened the pull request:
This PR adds a new implementation of the Export process that solves some of the problems the old process presented, especially when exporting a large number of submissions. The main problems were:
- Contingency issues on shared state that performed badly when exporting many forms
- Saturation of the EventBus with progress events

Thanks to the changes this new process introduces (mainly by using Java8 Streams), this also paves the way for follow-up PRs with error management and user feedback improvements.

The old export process has been preserved and it can be toggled with a new GUI configuration setting and CLI flag. The new export process is enabled by default unless users configure it otherwise.

#### What has been done to verify that this works as intended?
The commit https://github.com/opendatakit/briefcase/pull/457/commits/b4dc0b4b2cdf816ee96290a24ce1ef8017d6047d a comprehensive, outside-in test suite that compares outputs from the old and the new processes to verify that we haven't broken any feature:
- Basic exporting of plain and encrypted forms, with all supported data types
- Exporting forms with cascading and non-cascading repeat groups
- Exporting with a date range submission filter
- Exporting forms with media attachments, both plain and encrypted, covering various corner cases
- The overwrite files feature

    For more info on this, [check the test files](https://github.com/ggalmazor/briefcase/tree/improve_export_performance/test/java/org/opendatakit/briefcase/export) and [the form files they use](https://github.com/ggalmazor/briefcase/tree/improve_export_performance/test/resources/org/opendatakit/briefcase/export)

I have also tested manually the feature toggle on the GUI and the new CLI flag to run the old version of the export process.

#### Notes about performance

My machine is an i7 with 6 cores (12 threads), it has 32 GB of memory and I'm running the tests on an nvme hard drive.

Looking to the time it takes to export a form with 50K submissions, I get mixed results that I can't quite explain. In my machine, the old implementation consistently produces the result CSVs in ~17 seconds and the new algorithm dances around that value, sometimes being slower than that.

This is a list of changes I've tried to pinpoint a source for this inconsistent durations:
- Use the original submission file reading and sorting code to avoid any problem derived from using Java NIO2 or Streams for that.
- Reduce per-submission created number of objects by removing wrapping objects like Optionals and using primitives downstream, to reduce garbage collection problems
- Start Briefcase with a bigger minimum heap memory settings to reduce resizing during garbage collection
- Pass the OutputStreamWriter instances downstream to directly write in the files where any value is produced, instead of producing the values, collecting and writing them in the terminal operation.

None of this changes made the new implementation run consistently faster than the old one by themselves.

I'm curious about performance metrics on systems with fewer cores and memory. Any feedback will be very welcome.

[This is the dataset I've used](https://drive.google.com/open?id=1czxbJk3dIw3CY1AGlTlYozwyyKLNK_q8). You can decompress that into your Briefcase storage directory.


#### Why is this the best possible solution? Were any other approaches considered?
This is a total revamping of the export process. Progressively evolving the old one was considered, but it was too challenging to deal with a procedural style codebase which heavily relies on side effects to produce results.

Almost all the code of the new implementation started as a copy&paste of the old one, and then it was refactored using the outside-in tests as a baseline.

Going forward, there is a lot of room for design and performance improvements. The resulting code base should be more declarative and rely on immutable, side-effect free code as much as possible, which is easier to understand and refactor.

#### Are there any risks to merging this code? If so, what are they?
Even though the test suite has a really high coverage metrics both on the old and new processes, it is possible that some use case slipped our safety net. That's why we provide a feature toggle to reduce the impact of any issue, should this happen.

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
Nope.

dcbriccetti commented on src/org/opendatakit/briefcase/export/ExportProcessTracker.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  public void incAndReport() {
+    this.exported++;
```
You sure you want the “this.”?

dcbriccetti commented on src/org/opendatakit/briefcase/export/SubmissionMetaData.java:
```diff
@@ -0,0 +1,154 @@
+          .map(iso8601DateTime -> {
+            // Sometimes we can get just an ISO8601-ish string lacking of
+            // the minutes part of the offset block. Fix that.
+            return iso8601DateTime.length() - iso8601DateTime.lastIndexOf("+") == 3
```
This assumes everybody is *ahead* of UTC. The parse fails with `2018-01-15T00:45:36.127-08`, for example.

dcbriccetti commented on src/org/opendatakit/briefcase/export/SubmissionMetaData.java:
```diff
@@ -0,0 +1,154 @@
+            // the minutes part of the offset block. Fix that.
+            return iso8601DateTime.length() - iso8601DateTime.lastIndexOf("+") == 3
+                ? iso8601DateTime + ":00"
+                : iso8601DateTime;
```
That doesn’t work. How about this?
```
            return iso8601DateTime.length() - iso8601DateTime.lastIndexOf(":") == 3
                ? iso8601DateTime
                : iso8601DateTime + ":00";
```

dcbriccetti commented on src/org/opendatakit/briefcase/export/SubmissionMetaData.java:
```diff
@@ -0,0 +1,154 @@
+                ? iso8601DateTime + ":00"
+                : iso8601DateTime;
+          })
+          .map(OffsetDateTime::parse);
```
Do you need to catch DateTimeParseException from parse? How confident are you the date strings will be in the right format?

dcbriccetti replied:
This works better:
```
  Optional<OffsetDateTime> getSubmissionDate() {
    if (submissionDate == null)
      submissionDate = root.getAttributeValue("submissionDate")
          .map(SubmissionMetaData::regularizeDateTime)
          .map(OffsetDateTime::parse);
    return submissionDate;
  }

  /** Fixes ISO8601-ish strings not in this form: 2018-05-13T17:32:57±00:00 */
  private static String regularizeDateTime(String iso8601DateTime) {
    // 2018-04-26T08:58:20.525Z
    if (iso8601DateTime.endsWith("Z")) // Replace the Z with a zero offset
      return iso8601DateTime.substring(0, iso8601DateTime.length() - 1) + "+00:00";

    // 2018-05-13T17:32:57+00  Add the minutes if not present
    return iso8601DateTime + (iso8601DateTime.length() - iso8601DateTime.lastIndexOf(":") == 3 ? "" : ":00");
  }
```

dcbriccetti commented on src/org/opendatakit/briefcase/export/Model.java:
```diff
@@ -0,0 +1,252 @@
+  }
+
+  /**
+   * Method that lets iterate and (flat)map over all the children {@link Model} members of
```
Hey, let’s iterate! :-)
I suggest,
~~Method that lets~~ Iterates and (flat)maps over ...

dcbriccetti commented on src/org/opendatakit/briefcase/export/Model.java:
```diff
@@ -0,0 +1,252 @@
+  }
+
+  /**
+   * Method that lets iterate over all the children {@link Model} members of
```
~~Method that lets iterate~~ Iterates over

dcbriccetti commented on src/org/opendatakit/briefcase/export/XmlElement.java:
```diff
@@ -0,0 +1,306 @@
+  /**
+   * Returns the value inside this element.
+   * <p>
+   * If no value is found, this method will throw.
```
Throw requires an object. I suggest “heave” or “upchuck”, which don’t. Heh heh.

dcbriccetti commented on src/org/opendatakit/briefcase/export/SubmissionMetaData.java:
```diff
@@ -0,0 +1,154 @@
+ * generated while parsing submissions and they are used to read
+ * each value only once.
+ * <p>
+ * All its members are lazily evaluated to avoid unnecessary parsing
```
.

dcbriccetti commented on src/org/opendatakit/briefcase/export/XmlElement.java:
```diff
@@ -0,0 +1,306 @@
+  /**
+   * Returns the value inside this element.
+   * <p>
+   * If no value is found, this method will throw.
```
You should say what it will throw. Throw requires an object.

dcbriccetti replied:
>will throw an exception

Touché!

dcbriccetti commented:
I don’t understand why we’d want to keep the old export process.

dcbriccetti commented on src/org/opendatakit/briefcase/ui/export/ExportPanel.java:
```diff
@@ -19,6 +19,9 @@
 import static java.util.stream.Collectors.joining;
 import static java.util.stream.Collectors.toList;
 import static org.opendatakit.briefcase.export.ExportForms.buildCustomConfPrefix;
+import static org.opendatakit.briefcase.export.ExportOutcome.ALL_EXPORTED;
```
This is one of the special cases where I’d argue for the *. We’re using all three.

ggalmazor replied:
Yup. That would improve the code...

dcbriccetti commented:
Not part of this PR, but something you added in February. Do you get these helpful notices from IDEA? I usually resolve this one. Fewer fields is better.
![screen shot 2018-05-12 at 4 09 13 pm](https://user-images.githubusercontent.com/55099/39962292-2f071598-55ff-11e8-8cf5-f9609a7ab6a7.png)

dcbriccetti commented on src/org/opendatakit/briefcase/export/ExportToCsv.java:
```diff
@@ -0,0 +1,153 @@
+      EventBus.publish(ExportEvent.failure(formDef, "All submissions have been skipped"));
+
+    return exportOutcome;
+  }
```
If the above is your personal preference, fine, but you might consider 
```
switch (exportOutcome) {
  case ALL_EXPORTED:
    EventBus.publish(ExportEvent.successForm(formDef, (int) exportTracker.total));
    break;
  case SOME_SKIPPED:
    EventBus.publish(ExportEvent.partialSuccessForm(formDef, (int) exportTracker.exported, (int) exportTracker.total));
    break;
  case ALL_SKIPPED:
    EventBus.publish(ExportEvent.failure(formDef, "All submissions have been skipped"));
    break;
}
```
or even
```
Object event = null;
switch (exportOutcome) {
  case ALL_EXPORTED:
    event = ExportEvent.successForm(formDef, (int) exportTracker.total);
    break;
  case SOME_SKIPPED:
    event = ExportEvent.partialSuccessForm(formDef, (int) exportTracker.exported, (int) exportTracker.total);
    break;
  case ALL_SKIPPED:
    event = ExportEvent.failure(formDef, "All submissions have been skipped");
    break;
}
EventBus.publish(event);
```

ggalmazor replied:
But my if expression takes also into account the number of processed submissions... How can we express that in a `switch`?

dcbriccetti replied:
The switch on the enum with no default makes it explicit that we are addressing all three mutually exclusive conditions. For what that’s worth.

dcbriccetti replied:
Using else/if’s would get you partway there. 
I don’t care much about this one.

dcbriccetti commented on src/org/opendatakit/briefcase/export/CipherFactory.java:
```diff
@@ -0,0 +1,119 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handles the initialization of Cipher objects for the decryption of submission
```
`{@link Cipher}`, to be consistent with the `@link`s in doc comments below.

dcbriccetti commented on src/org/opendatakit/briefcase/export/CipherFactory.java:
```diff
@@ -0,0 +1,119 @@
+  Cipher next() {
+    try {
+      ++ivSeedArray[ivCounter % ivSeedArray.length];
+      ++ivCounter;
```
I’m assuming you didn’t write
`++ivSeedArray[ivCounter++ % ...` because you thought it would be too hard to understand. Some don’t know the difference between pre- and post-increment?

ggalmazor replied:
Most of the code in this class is a mystery to me... :/

dcbriccetti replied:
Sorry, I thought you wrote that.

yanokwa commented:
@dcbriccetti The idea behind keeping the old export was for safety and give people a chance to use the other export if they were getting unexpected results. I guess they can always use the old binary...

dcbriccetti commented:
I see. I suggest we remove it. It complicates the review (a little), and means it has to be tested and maintained (and maybe even deprecated) going forward.

dcbriccetti commented on src/org/opendatakit/briefcase/export/CipherFactory.java:
```diff
@@ -0,0 +1,119 @@
+  private final byte[] ivSeedArray;
+  private int ivCounter = 0;
+
+  private CipherFactory(String instanceId, byte[] symmetricKeyBytes) throws CryptoException {
```
I don’t see consistency in when you use the `throws CryptoException` clause among the methods in this class with throw it.

ggalmazor replied:
That's a mistake! Fixing it! :)

dcbriccetti commented on src/org/opendatakit/briefcase/export/CipherFactory.java:
```diff
@@ -0,0 +1,119 @@
+   * Factory that initializes a new {@link CipherFactory} for the given instance ID,
+   * encryption key and private key values.
+   *
+   * @throws CryptoException
```
These are missing the description (under what circumstances the exception would be thrown).
https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javadoc.html#throws

dcbriccetti commented on src/org/opendatakit/briefcase/export/CipherFactory.java:
```diff
@@ -0,0 +1,119 @@
+  private static final int IV_BYTE_LENGTH = 16;
+  private final SecretKeySpec symmetricKey;
+  private final byte[] ivSeedArray;
+  private int ivCounter = 0;
```
It’s not clear what “iv” is

dcbriccetti commented on src/org/opendatakit/briefcase/export/CipherFactory.java:
```diff
@@ -0,0 +1,119 @@
+      ivSeedArray = new byte[IV_BYTE_LENGTH];
+      for (int i = 0; i < IV_BYTE_LENGTH; ++i) {
+        ivSeedArray[i] = messageDigest[(i % messageDigest.length)];
+      }
```
Might be good to explain why we create the seed array this way.

ggalmazor replied:
That would help a lot, but I honestly don't know why this is done that way :P

dcbriccetti replied:
Let’s forget that, then. And anywhere else I thought you wrote it and then I asked for something.

dcbriccetti commented on src/org/opendatakit/briefcase/export/CipherFactory.java:
```diff
@@ -0,0 +1,119 @@
+      md.update(instanceId.getBytes("UTF-8"));
+      md.update(symmetricKeyBytes);
+      byte[] messageDigest = md.digest();
+      ivSeedArray = new byte[IV_BYTE_LENGTH];
```
You may have explained before why you do it this way, but why not initialize in the field declaration?

ggalmazor replied:
This is just a copy of `org.opendatakit.briefcase.util.CipherFactory`. I honestly don't know why encryption is dealt with this way... I'll rescue some comments from the original, just in case.

dcbriccetti commented on src/org/opendatakit/briefcase/export/CsvMapper.java:
```diff
@@ -0,0 +1,334 @@
+    values.add(Pair.of(element.fqn() + "-Longitude", Optional.ofNullable(fields[1]).orElse(null)));
+    values.add(Pair.of(element.fqn() + "-Altitude", Optional.ofNullable(fields[2]).orElse(null)));
+    values.add(Pair.of(element.fqn() + "-Accuracy", Optional.ofNullable(fields[3]).orElse(null)));
+    return values;
```
I didn’t like the duplication here so I tried my hand at Java 8 streams. Would you consider something like this?
```
    String[] tags = new String[] {"Latitude", "Longitude", "Altitude", "Accuracy"};
    return IntStream.range(0, tags.length).mapToObj(i ->
        Pair.of(element.fqn() + "-" + tags[i], Optional.ofNullable(fields[i]).orElse(null))
    ).collect(Collectors.toList());
```
![screen shot 2018-05-12 at 5 33 41 pm](https://user-images.githubusercontent.com/55099/39962664-b16c8b66-560a-11e8-8d60-391aca685dc0.png)

dcbriccetti replied:
Thanks for that, and for leading us into the past future with Java 8 functional features

dcbriccetti commented on src/org/opendatakit/briefcase/export/DataType.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * This enum encodes JavaRosa data types as described in {@link org.javarosa.core.model.Constants}
+ */
+enum DataType {
```
I think we should declare this enum in JavaRosa, even if we don’t convert JR code to use the enum right away. I’ll make such a PR if I don’t hear disagreement.

ggalmazor replied:
That's great, thanks!

ggalmazor replied:
Has the a new version of JR including this enum been released yet?

lognaturel replied:
Yes! Still in snapshot but you can use it. The version is 2.10.0-SNAPHOT

ggalmazor replied:
Ugh... possibly newbie question here, sorry...

I'm getting an error after updating the JR version in the build file:
```
   > Could not find org.opendatakit:opendatakit-javarosa:2.10.0-SNAPHOT.
     Searched in the following locations:
         https://repo.maven.apache.org/maven2/org/opendatakit/opendatakit-javarosa/2.10.0-SNAPHOT/opendatakit-javarosa-2.10.0-SNAPHOT.pom
         https://repo.maven.apache.org/maven2/org/opendatakit/opendatakit-javarosa/2.10.0-SNAPHOT/opendatakit-javarosa-2.10.0-SNAPHOT.jar
         https://oss.sonatype.org/content/groups/public/org/opendatakit/opendatakit-javarosa/2.10.0-SNAPHOT/opendatakit-javarosa-2.10.0-SNAPHOT.pom
         https://oss.sonatype.org/content/groups/public/org/opendatakit/opendatakit-javarosa/2.10.0-SNAPHOT/opendatakit-javarosa-2.10.0-SNAPHOT.jar
```

lognaturel replied:
🤔Really? I just tried 
```
   compile(group: 'org.opendatakit', name: 'opendatakit-javarosa', version: '2.10.0-SNAPSHOT') {
        exclude group: 'org.slf4j'
    }
```
and it worked. Maybe some kind of caching? Clean and build again?

You will need to exclude slf4j but that's not what the error was about.

ggalmazor replied:
Thanks! That did the trick :)

Why exclude slf4j, though?

lognaturel replied:
I believe that if the versions in JR and Briefcase get out of sync the build will fail if it's included transitively from JR. They match now so I suppose we can leave it in and make sure to keep the versions in sync. I'd tend to exclude it so we don't have to deal with cryptic errors in the future.

dcbriccetti commented on src/org/opendatakit/briefcase/export/ExportEvent.java:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  public static ExportEvent end(FormDefinition form, long exported) {
+    return new ExportEvent(form.getFormId(), String.format("Export operation has ended", exported));
```
Gotta love IDEA.
![screen shot 2018-05-12 at 9 15 09 pm](https://user-images.githubusercontent.com/55099/39963763-a6f725f0-5629-11e8-8f7b-55f7a4ad8314.png)

dcbriccetti commented on src/org/opendatakit/briefcase/reused/UncheckedFiles.java:
```diff
@@ -0,0 +1,271 @@
+
+  public static boolean isInstanceDir(Path dir) {
+    return Files.isDirectory(dir)
+        && dir.getFileName().toString().matches("^uuid[\\w]{8}-[\\w]{4}-[\\w]{4}-[\\w]{4}-[\\w]{12}$")
```
This is a little scary. Will it break if the naming convention changes? Should this test reside closer to where the convention is implemented?
It would be helpful to at least add a comment with an example of an instance directory so readers can match it up with the regex if they want to.

ggalmazor replied:
> This is a little scary. Will it break if the naming convention changes? Should this test reside closer to where the convention is implemented?

That's a very good point! I'll add a comment and we can open an issue to fix this design flaw after we get this merged. This affects the pulling process where these directories are created

dcbriccetti replied:
That’s a good comment.

dcbriccetti commented on src/org/opendatakit/briefcase/reused/UncheckedFiles.java:
```diff
@@ -0,0 +1,271 @@
+        && Files.exists(dir.resolve("submission.xml"));
+  }
+
+  public static boolean isSubmissionFile(Path path) {
```
Is this method called?

dcbriccetti commented on src/org/opendatakit/briefcase/export/FieldMapper.java:
```diff
@@ -0,0 +1,63 @@
+
+  /**
+   * {@link FieldMapper} factory that will just apply some concrete {@link XmlElement} mapper
+   * assuming that there is a fixed amount of key-value pairs to be returned.
```
Sometimes the amount/number distinction is not worth fussing about, but here I think you should say “fixed number” instead.
https://www.google.com/search?q=amount+number

dcbriccetti commented on src/org/opendatakit/briefcase/export/CsvMapper.java:
```diff
@@ -0,0 +1,334 @@
+  private static List<Pair<String, String>> formattedDate(XmlElement element, DateFormat formatter) {
+    String value = element.getValue();
+    return singletonList(Pair.of(element.fqn(), value.isEmpty() ? null : formatter.format(parseDate(value))));
+  }
```
This fails when the date tag has no body. Here’s a guess about fixing it, but maybe you have some better chaining construct, or a different solution altogether.
![screen shot 2018-05-12 at 10 01 07 pm](https://user-images.githubusercontent.com/55099/39964047-3c2c0086-5630-11e8-9d87-2d217936ca3b.png)

dcbriccetti replied:
Oh, I see for binary you do this:
```
    if (!element.hasValue())
      return empty(element.fqn());
```

ggalmazor replied:
I see that it's not consistent how I deal with empty elements through the code at this point... Let's discuss how it should behave when dealing with empty elements and then I'll change this commit.

We need to discuss these data types:

| Data type | Expected output |
| --- | --- |
| text (generic field) | an empty string |
| date & time types | 1 empty string |
| geopoint | 4 empty strings |
| binary | 1 empty string |
| null (non repeating group) | should throw? |
| null (repeating group) | should throw? |

dcbriccetti replied:
I don’t know. Can you just do what the previous export did?

dcbriccetti commented on src/org/opendatakit/briefcase/export/CsvMapper.java:
```diff
@@ -0,0 +1,334 @@
+  private static List<Pair<String, String>> geopoint(XmlElement element) {
+    String[] fields = Optional.ofNullable(element.getValue())
+        .map(v -> v.split(" "))
+        .orElseThrow(() -> new BriefcaseException("No value on geopoint field"));
```
This also throws an exception if element is empty, presumably not completing the export.

dcbriccetti commented on src/org/opendatakit/briefcase/export/ExportEvent.java:
```diff
@@ -0,0 +1,65 @@
+    this.statusLine = statusLine;
+  }
+
+  public static ExportEvent progress(FormDefinition form, double percentage) {
```
I recommend some changes here to improve the presentation of these messages.

![screen shot 2018-05-13 at 10 41 48 am](https://user-images.githubusercontent.com/55099/39970091-765c8e32-569a-11e8-8fcf-28a0bd578355.png)

Let’s pluralize correctly. 

“Export operation” is longer and no better than “Export”.  Same for “Successfully exported” and “exported”, and the latter allows the whole message to fit in the GUI on my computer, as you can see below.

![screen shot 2018-05-13 at 10 40 40 am](https://user-images.githubusercontent.com/55099/39970105-c97121dc-569a-11e8-90fd-4776d85549a8.png)

I’m confused by the “Not exported yet”.

dcbriccetti replied:
Both of these statements can’t be true:
![screen shot 2018-05-17 at 1 41 11 pm](https://user-images.githubusercontent.com/55099/40202833-1eba4872-59d8-11e8-8e6f-6d4e7a8dd199.png)

dcbriccetti commented:
Here’s my branch off your branch where I’ve tried out the changes I am recommending. Help yourself. (Just wondering what “help yourself” sounds like to people not familiar with the idiom.)

dcbriccetti commented on src/org/opendatakit/briefcase/export/CipherFactory.java:
```diff
@@ -0,0 +1,119 @@
+      pkCipher.init(Cipher.DECRYPT_MODE, privateKey);
+      byte[] encryptedSymmetricKey = Base64.decodeBase64(base64EncryptedKey);
+      byte[] decryptedKey = pkCipher.doFinal(encryptedSymmetricKey);
+      return new CipherFactory(instanceId, decryptedKey);
```
11:38:52.334 [main] ERROR org.opendatakit.briefcase.util.FileSystemUtils - Error decrypting:1524644764247.jpg.enc
java.security.InvalidKeyException: Illegal key size
⋮
	at org.opendatakit.briefcase.util.CipherFactory.getCipher(CipherFactory.java:78)
	at org.opendatakit.briefcase.util.EncryptionInformation.getCipher(EncryptionInformation.java:59)
	at org.opendatakit.briefcase.util.FileSystemUtils.decryptFile(FileSystemUtils.java:455)

ggalmazor replied:
Can you confirm that you're using Java 8u161 or greater?

dcbriccetti replied:
JRE: 1.8.0_152-release-1136-b20 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.12.6

ggalmazor replied:
OK, with that JVM you should follow these complicated and boring instructions here: https://docs.opendatakit.org/encrypted-forms/#configuration

**OR**

Install JVM 1.8.0u161 or a greater version :)

dcbriccetti commented on test/java/org/opendatakit/briefcase/export/ExportToCsvScenario.java:
```diff
@@ -0,0 +1,295 @@
+              .filter(file -> !file.getFileName().toString().endsWith(".pem"))
+              .filter(file -> !file.getFileName().toString().endsWith(".tpl"))
+              .forEach(file -> {
+                Path target = submissionDir.resolve(file.getFileName().toString().substring(formName.length() + 1));
```
I find this more readable. You?
```
.filter(file -> {
  String name = file.getFileName().toString();
  return name.startsWith(formName + "_") && !name.endsWith(".pem") && !name.endsWith(".tpl");
}).forEach(file -> {
```

ggalmazor replied:
I'd prefer to leave them separate. I'll give those predicates a name to see if it's more clear that way.

dcbriccetti replied:
OK

dcbriccetti replied:
I like that less. The separate function is cool, but why not pass the extension in as an argument?

dcbriccetti commented on test/java/org/opendatakit/briefcase/export/ExportToCsvScenario.java:
```diff
@@ -0,0 +1,295 @@
+  }
+
+  private static String readInstanceId(Path path) {
+    return readAllLines(path).get(0).split("instanceID=\"")[1].split("\"")[0];
```
Keep in mind that some forms are gigantic. Why not just read the first line?
```
    return readOneLine(path).split("instanceID=\"")[1].split("\"")[0];
```
![screen shot 2018-05-13 at 1 03 45 pm](https://user-images.githubusercontent.com/55099/39971343-1eaf5926-56ae-11e8-8ec7-8087b3b0e4d5.png)

dcbriccetti commented on src/org/opendatakit/briefcase/export/SubmissionParser.java:
```diff
@@ -0,0 +1,187 @@
+        // are not contained in the given date range, but parsing is expensive
+        // enough to make it all in one pass and use Optionals to account
+        // for that
+        .map(path -> parseSubmission(path.resolve("submission.xml"), dateRange, isEncrypted, privateKey))
```
I had an error here:

java.io.UncheckedIOException: java.nio.file.FileSystemException: /private/tmp/ODK Briefcase Storage/forms/Independent Monitoring/instances/.DS_Store/submission.xml: Not a directory

Here is one solution:
```
        // for that
        .filter(path -> !path.toFile().isHidden())
        .map(path -> parseSubmission(path.resolve("submission.xml"), dateRange, isEncrypted, privateKey))
```

ggalmazor replied:
Something doesn't seem right... Could you copy the full trace here, please?

I don't understand how you're getting that error there and what has to do with `.DS_Store` beind hidden...

dcbriccetti replied:
Browse your data with Finder (are you using MacOS?) and it will create the .DS_Store directories.

dcbriccetti commented:
Regarding performance, I can send you a YourKit snapshot if you like. We have a license, if you weren’t aware. Here’s my little system:

  Model Name:	MacBook Pro
  Model Identifier:	MacBookPro11,3
  Processor Name:	Intel Core i7
  Processor Speed:	2.8 GHz
  Number of Processors:	1
  Total Number of Cores:	4
  L2 Cache (per Core):	256 KB
  L3 Cache:	6 MB
  Memory:	16 GB

ggalmazor commented:
Thanks, @dcbriccetti, it's an impressive review! I'm trying to limit my brain bandwidth to one hard task at a time and I'm currently finishing some other stuff. I hope I can retake this PR during the next days and process your review.

ggalmazor commented on src/org/opendatakit/briefcase/export/ExportProcessTracker.java:
```diff
@@ -0,0 +1,52 @@
+    if (percentage > lastReportedPercentage && percentage % 10 == 0) {
+      EventBus.publish(ExportEvent.progress(form, percentage.doubleValue() / 100));
+      lastReportedPercentage = percentage;
+    }
```
I think it won't produce the same values because dividing `int / double` produces an `int` by truncating decimals.

dcbriccetti replied:
scala> 5 / 2
res0: Int = 2

scala> 5 / 2D
res1: Double = 2.5

ggalmazor replied:
Ooh, alright!

ggalmazor commented:
> (About the old implementation) I see. I suggest we remove it. It complicates the review (a little), and would have to be tested and maintained (and maybe even deprecated) going forward.

The idea is to remove it after some time (maybe a pair of releases) after the new one is a little more battle tested.

dcbriccetti commented:
>The idea is to remove it after some time (maybe a pair of releases) after the new one is a little more battle tested.

As Yaw said, users can change to the previous version of Briefcase if they have problems.

I don’t care very much about this.

ggalmazor commented:
I don't have a strong opinion about this. If @yanokwa changes his opinion we can remove the old one.

dcbriccetti commented:
>I don't have a strong opinion about this. If @yanokwa changes his opinion we can remove the old one.

You mean if @yanokwa  _maintains_ his position? He seemed to agree that including both wasn’t needed.

yanokwa commented:
I discussed the removal with @lognaturel and she doesn't feel strongly about it. It is very easy to roll back and this saves us some maintenance, so let's do what @dcbriccetti suggested and remove the crutch, @ggalmazor!

ggalmazor commented:
OK, now there no place outside the test where the old implementation is used.

Next, I'll record the expected outputs to convert the GoldenMaster tests in place into QA tests that don't exercise the old implementation.

ggalmazor commented:
I've converted the tests but now I'm having some problems with time zones... I'll continue tomorrow :P

ggalmazor commented:
@yanokwa, @dcbriccetti, I think it's ready now.

dcbriccetti commented on src/org/opendatakit/briefcase/export/Model.java:
```diff
@@ -0,0 +1,253 @@
+  }
+
+  /**
+   * Iterates the children of this instance and returns the flatmapped result of mapping
```
I think it needs to be “iterate over the children”.
https://english.stackexchange.com/questions/23023/iterate-vs-iterate-over

dcbriccetti commented:
I’m glad we agreed on so many of my suggested changes. What’s left to resolve?
- empty tags
- key error
- not exported yet
- hidden folders

ggalmazor commented:
Thanks, @dcbriccetti ! It's a lifesaver to have you onboard for these huge changes :)

**empty tags**

I'll throw in a new test focused on testing this to verify it works as expected.

**key error**

My best bet for this one is that you're running Briefcase with a JVM older than 1.8.0_161. Can you check that? Previous versions need additional config to work with the cryptographic library we're using.

I can recommend http://sdkman.io to manage JVM and other SDK versions in your computer.

**not exported yet**

This is used when the form hasn't been exported by the user yet. After the first export, the last export date is shown there.

Is there something you would change here?

**hidden folders**

I still don't understand how that trace is related to having that hidden folder. Could you check the contents of that `submissions.xml` file? is that actually a valid submission?

Anyway, I'll add a filter that excludes any directory that doesn't look like a legit instance directory.

ggalmazor commented:
Regarding empty tags, this is what the old implementation will produce:

| Type               | Output            | Example                   | 
| ------------------ | ----------------- | ------------------------- |
| string             | `""`              | `something,"",something`  |
| int                | `""`              | `something,"",something`  |
| decimal            | `""`              | `something,"",something`  |
| date               | empty column      | `something,,something`    |
| time               | empty column      | `something,,something`    |
| date_time          | empty column      | `something,,something`    |
| geopoint           | 4 x empty columns | `something,,,,,something` |
| geotrace           | `""`              | `something,"",something`  |
| geoshape           | `""`              | `something,"",something`  |
| barcode            | `""`              | `something,"",something`  |

Does this seem right to you, @lognaturel, @yanokwa, @dcbriccetti? I think numeric types should be encoded without double quotes...

ggalmazor commented:
I've fixed how we handle empty tags. Both `<field></field>` and `<field/>` should work as defined on the table from my last comment.

I've also added the filter that will process only submissions inside correct instance folders.

dcbriccetti commented:
>I still don't understand how that trace is related to having that hidden folder. 

map(path -> parseSubmission(path.resolve("submission.xml"), dateRange, isEncrypted, privateKey))

You were trying to open "submission.xml" in every directory in `instances`. There is no such file in the .DS_store directories.
```
dbl:/tmp/ODK Briefcase Storage/forms/All widgets/instances $ ls -a
.                                        .DS_Store
..                                       uuide8042ecc-9966-4d6f-9db3-2a5b2af3d96d
```

You’ve solved it with your new filter.

dcbriccetti commented on src/org/opendatakit/briefcase/export/CsvMapper.java:
```diff
@@ -0,0 +1,352 @@
+      default:
+        return false;
+    }
+  }
```
I find this much simpler. You?

![screen shot 2018-05-17 at 2 17 58 pm](https://user-images.githubusercontent.com/55099/40204388-321cd844-59dd-11e8-9204-266a63f651f8.png)

ggalmazor commented:
> You were trying to open "submission.xml" in every directory in instances. There is no such file in the .DS_store directories.

But the trace said "Not a directory"... I agree that now it's solved... Let's move on :P

dcbriccetti commented:
>But the trace said "Not a directory"... I agree that now it's solved... Let's move on :P

That puzzled me, too. I think the error message is misleading.

dcbriccetti commented:
Remaining items:

me: key error
@yanokwa @lognaturel: verify correct handling of nulls
@ggalmazor: contradictory “not exported yet”

What else?

ggalmazor commented:
@dcbriccetti, @yanokwa, @lognaturel, I think there is nothing pending to be changed now.

We could discuss some incoherence I've found on how the old implementation dealt with empty tags - see https://github.com/opendatakit/briefcase/pull/457#issuecomment-389788696 - but, at least, we can be sure that the new implementation does just the same. We could change that on further PRs if we want.

yanokwa commented:
Agreed with @ggalmazor that we should file an issue for the empty values and discuss that there. Matching the old implementation makes it good enough to test.

@ggalmazor/dcbriccetti can either of you explain the issue with the contradictory 'not exported yet'. Explain Like I'm Five.

ggalmazor commented:
> can either of you explain the issue with the contradictory 'not exported yet'. Explain Like I'm Five.

@dcbriccetti detected a problem on the UI: the column for "last export date" wasn't being updated after exporting a form, therefore, the status was telling "Successfully exported" and the last export date was "Not exported yet", which is contradictory.

What was happening was that I changed some code related to events and subscriptions and forgot to adapt one method in `ExportForms.appendStatus()` responsible for updating the last export date when we detected that a form had been successfully exported.

Now that the method works correctly, the last export date is updated and the UI doesn't show contradictory information anymore.

ggalmazor commented:
Thanks for all your feedback, @yanokwa, and @dcbriccetti! I know that reviewing these monster PRs requires a huge effort.

kkrawczyk123 commented:
My efficiency results are not very impressive. On my ubuntu machine export of 50k submissions takes 1 minute, I see “Export has started” immediately, then % of progress in Export Status column and then amount of exported submissions. (I have 16GB of ram memory there). The same activity looks terrible on my Mac mini (8gb of ram) and takes like forever on my Windows VM (4gb of ram). I wasn’t even able to notice any % status of progress, and on Windows it took like 5 minutes for “Export has started” to show up. And it started, started and froze my briefcase ;( so after 30 minutes of waiting export still didn’t finish. I have some file in my export directory after nearly an hour I have 31800 submissions in that file.

I have noticed that progress bar is visible only for few seconds then it disappears (On all three systems right after Export has started appears in Export status on all 3 systems and when bulk export also) Is this an intended behavior? 

I have already mentioned it when progress bar was added when you do not add pem key file location when export encrypted form error message shows, export doesn’t start but the progress bar is active and it doesn’t want to stop to dance. (all three systems) 

@opendatakit-bot unlabel "needs testing"

yanokwa commented:
@kkrawczyk123 when it froze on Windows, did you get any error messages? Does the behavior change if you try the export on the CLI?

ggalmazor commented:
I know this is a pain to test, but it would be great to know how the master branch deals with the same amount of data.

ggalmazor commented:
Also, the expected behavior would be that the UI doesn't block at any time, the progress bar is active all the time and we get progress update every 10%. 

For long exports like these, it could be better to get progress updates each 1%, maybe.

yanokwa commented:
@kkrawczyk123 is it easy for you to try the efficiency test on master, so we can compare?

ggalmazor commented:
> I have already mentioned it when progress bar was added when you do not add pem key file location when export encrypted form error message shows, export doesn’t start but the progress bar is active and it doesn’t want to stop to dance. (all three systems)

That's a bug! I'm on it :)

ggalmazor commented:
@kkrawczyk123, I've fixed the error with the PEM file you have reported. I've also reviewed and improved how the UI locks itself while exporting to prevent user interaction while an export is happening.

I've also made some small changes to how the export process is launched which could improve the situation with lots of submissions.

kkrawczyk123 commented:
## Branch vs. Master

### Round 1st:
Export of 50k submissions.

**Branch Results:**
_Ubuntu:_ less than 1 minute
_MacOs_: after 17 minutes Unexpected error, exported submissions 40k
![screenshot from 2018-05-29 12-01-18](https://user-images.githubusercontent.com/33342488/40652577-da0efc12-6338-11e8-928d-96e451f8a95b.png)

_Windows:_ the same error message as on MacOS after 15 minutes, exported submissions: 29k
![screenshot from 2018-05-29 10-18-14](https://user-images.githubusercontent.com/33342488/40652652-1ef1e132-6339-11e8-8826-d232693fd701.png)

**Master Results:**
_Ubuntu:_ the same results as on the branch, less than minute
_MacOS:_ success after 8 minutes
_Windows:_ success after 15 mintes

### Round 2nd:
Bulk export of 50k submissions and other forms and submissions.

**Branch Results:**
_Ubuntu_: 1 minute 
_MacOS:_ the same error message after 5 minutes
_Windows_: the same error message after 8 minutes

**Master Results:**
_Ubuntu:_ same as on the branch, like 1 minute
_MacOS:_ success after 8 minutes
_Windows:_ success after 20 minutes
(amount of forms could be different on all platforms but still, success is a success)

### Round 3rd:
Cmd export.

**Branch Results:**
_Ubuntu:_ the same result in time meaning but prefix -f didn't worked
_MacOS:_ process didn't fnished, error in cmd response
_Windows_: process didn't fnished, error in cmd response
Content of cmd response:
10:52:52.822 [main] ERROR org.opendatakit.common.cli.Cli - Error
java.lang.OutOfMemoryError: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.lang.reflect.Constructor.newInstance(Unknown Source)
        at java.util.concurrent.ForkJoinTask.getThrowableException(Unknown Source)
        at java.util.concurrent.ForkJoinTask.reportException(Unknown Source)
        at java.util.concurrent.ForkJoinTask.invoke(Unknown Source)
        at java.util.stream.Nodes.collect(Unknown Source)
        at java.util.stream.ReferencePipeline.evaluateToNode(Unknown Source)
        at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
        at java.util.stream.SortedOps$OfRef.opEvaluateParallel(Unknown Source)
        at java.util.stream.AbstractPipeline.opEvaluateParallelLazy(Unknown Source)
        at java.util.stream.AbstractPipeline.sourceSpliterator(Unknown Source)
        at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
        at java.util.stream.ReferencePipeline.forEachOrdered(Unknown Source)
        at org.opendatakit.briefcase.export.ExportToCsv.export(ExportToCsv.java:114)
        at org.opendatakit.briefcase.operations.Export.export(Export.java:88)
        at org.opendatakit.briefcase.operations.Export.lambda$static$2(Export.java:55)
        at org.opendatakit.common.cli.Cli.lambda$run$4(Cli.java:126)
        at java.lang.Iterable.forEach(Unknown Source)
        at org.opendatakit.common.cli.Cli.run(Cli.java:123)
        at org.opendatakit.briefcase.Launcher.main(Launcher.java:79)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at org.eclipse.jdt.internal.jarinjarloader.JarRsrcLoader.main(JarRsrcLoader.java:58)
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
        at java.util.stream.StreamSupport.stream(Unknown Source)
        at java.util.stream.Stream.concat(Unknown Source)
        at org.opendatakit.briefcase.export.XmlElement.lambda$flatten$6(XmlElement.java:227)
        at org.opendatakit.briefcase.export.XmlElement$$Lambda$120/1529060733.apply(Unknown Source)
        at java.util.stream.ReferencePipeline$7$1.accept(Unknown Source)
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
        at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
        at java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(Unknown Source)
        at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Unknown Source)
        at java.util.stream.ReferencePipeline$Head.forEach(Unknown Source)
        at java.util.stream.ReferencePipeline$7$1.accept(Unknown Source)
        at java.util.ArrayList$ArrayListSpliterator.tryAdvance(Unknown Source)
        at java.util.stream.ReferencePipeline.forEachWithCancel(Unknown Source)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(Unknown Source)
        at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
        at java.util.stream.FindOps$FindOp.evaluateSequential(Unknown Source)
        at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
        at java.util.stream.ReferencePipeline.findFirst(Unknown Source)
        at org.opendatakit.briefcase.export.XmlElement.findElement(XmlElement.java:113)
        at org.opendatakit.briefcase.export.CsvMapper.lambda$null$12(CsvMapper.java:194)
        at org.opendatakit.briefcase.export.CsvMapper$$Lambda$141/1536031937.accept(Unknown Source)
        at java.util.ArrayList.forEach(Unknown Source)
        at org.opendatakit.briefcase.export.Model.forEach(Model.java:72)
        at org.opendatakit.briefcase.export.CsvMapper.lambda$getRepeatCsvLine$13(CsvMapper.java:190)
        at org.opendatakit.briefcase.export.CsvMapper$$Lambda$139/461160828.accept(Unknown Source)
        at java.util.ArrayList.forEach(Unknown Source)
        at org.opendatakit.briefcase.export.CsvMapper.getRepeatCsvLine(CsvMapper.java:187)
        at org.opendatakit.briefcase.export.ExportToCsv.lambda$null$3(ExportToCsv.java:102)
        at org.opendatakit.briefcase.export.ExportToCsv$$Lambda$131/371439501.accept(Unknown Source)
        at java.util.ArrayList.forEach(Unknown Source)

**Master Results:**
_Ubuntu_: the same result in time meaning and prefix -f worked very well
_MacOS:_ export succeeded in 2 minutes, prefix -f worked very well
_Windows:_ export succeeded in 7 minutes, prefix -f worked very well


I confirm that bug connected with pem key file has been fixed and also that UI look while export is ongoing has been improved :)

kkrawczyk123 commented:
One more UI think that I've noticed and forgot about in the previous comment, on Ubuntu and Windows scroll overlaps a bit details buttons on MacOS everything looks good, screenshots of ubuntu and windows attached: 
![exportwin](https://user-images.githubusercontent.com/33342488/40654681-42175786-633f-11e8-8b11-173414960870.png)
![export](https://user-images.githubusercontent.com/33342488/40654682-4234c1f4-633f-11e8-9bd1-a113b4c8ff69.png)

ggalmazor commented:
Thanks, @kkrawczyk123. This information is very valuable! :)

ggalmazor commented:
We knew that the new code requires more memory to process everything. In my setup, it was approximately a 125% of the memory that the master branch needs.

It shouldn't take as much as 8GB of memory... Could you maybe close running programs and see how much free memory you have on your mac before starting and try again the new branch version?

kkrawczyk123 commented:
Briefcase was the only running program on Mac when I was testing  - I use it only in this project so I could have opened only browser and some folders but it shouldn't have a significant impact on that.

ggalmazor commented:
Great :) then I guess we only would have to know how much free memory is left before starting the export.

ggalmazor commented:
@kkrawczyk123 regarding the scroll problem, I think it should be filed as a separate issue (we haven't changed the UI layout on this PR). Could you create it, please?

kkrawczyk123 commented:
As u wish I will report :)

ggalmazor commented:
@kkrawczyk123, I've improved the memory usage... Could you test this PR again, please?

ggalmazor commented:
@opendatakit-bot label "needs testing"

kkrawczyk123 commented:
I was able to export with success on all 3 systems, results:

### Ubuntu:
**export:** less than 1 minute
**cmd export:** less than 1 minute

### MacOS:
**export:** 11 minutes
**cmd export:** 1 minute

### Windows:
**export:** 22 minutes
**cmd export:** 12 minutes.

The only  one issue that I've noticed is that -f prefix doesn't work on this pr on on 3 systems when exportig using cmd. 

@opendatakit-bot unlabel "needs testing"

ggalmazor commented:
Thanks, @kkrawczyk123! I'll fix the `-f` for next monday. I think I removed it by accident since the filename can't be set from the UI.

dcbriccetti commented:
![briefcase in 100 mb](https://user-images.githubusercontent.com/55099/40888893-dc3ea784-6712-11e8-90ed-75aa8da93efe.png)

This code now runs with a max heap of only 100 MB. Nice work! There are other optimizations that we could make (not parsing each form twice), but I don’t think they need to be made now.

ggalmazor commented:
@kkrawczyk123, I've fixed the `-f` cli param :)

kkrawczyk123 commented:
I confirm.
@opendatakit-bot label "behavior verified"
